# Python Exercise 8 - Homework 1


### This exercise is part you the homework you need to submit for your final grade

* Please try to answer all the questions in this exercise and send your solution to me via email before the end of April 18th 2017.
* This homework is worth 15% of your final grade
* Please prepare and submit your Exercise according to the instructions below. **I will deduce points** if your exercise is not submitted according to the instructions or is submitted late.
* For each question you should provide source to answer it and summarize your findings (in Chinese or English, in any case I will not grade your English as long as I can understand the meaning).

### Instructions for submission
* Enter your name, major and your student number in the field below (in Chinese). Double click on the field to edit it.

**Name:**  项闰冶 **major** 计算机科学与技术 **student number** 140104200230   

* Change the name of the notebook to the following format
```
Python HW 1 <yourname> <your student no>
```
(you can change the name in the menu below `File->Rename`. So, if you name is 马小龙 and your student number is 2018101111000111, you should change the name of the notebook to `Python HW 1 马小龙 2018101111000111`

* Export the notebook both as a HTML file and as notebook (choose `File->Download as` from the menu)
* Send both files to me via email to 2485707749@qq.com before the end of Tuesday, April 18th 2017
* your email should mention both your name and your student number in the subject
* to repeat, your name and student number should appear in three different places: in the text above, in the filename, and in the subject of your email!


**I will deduce points if your exercise is not submitted according to the instructions or is submitted late.**



### About the exercise

In this exercise we will use data from the San Francisco crime statistics from Summer 2014. You will
need the file `sanfrancisco_incidents_summer_2014.csv`, please copy it into the same folder as this notebook.


The file is not particularly large (5.9MB), but some Desktop programs might already have problems with it. Let's first open the file into a pandas DataFrame:

In [1]:
import numpy as np
import pandas as pd
df = pd.read_csv('sanfrancisco_incidents_summer_2014.csv')
df

IncidntNum                     Category  \
0       140734311                        ARSON   
1       140736317                 NON-CRIMINAL   
2       146177923                LARCENY/THEFT   
3       146177531                LARCENY/THEFT   
4       140734220                 NON-CRIMINAL   
5       140734349                DRUG/NARCOTIC   
6       140734349                DRUG/NARCOTIC   
7       140734349  DRIVING UNDER THE INFLUENCE   
8       140738147               OTHER OFFENSES   
9       140734258                     TRESPASS   
10      140738711                VEHICLE THEFT   
11      146178410                LARCENY/THEFT   
12      140787570                 NON-CRIMINAL   
13      140735604                VEHICLE THEFT   
14      140738147                VEHICLE THEFT   
15      140734145                      ASSAULT   
16      140734214                        FRAUD   
17      140734214                      ASSAULT   
18      140734098               SUSPICIOUS OCC   
19      140734098                 NON-CRIMINAL   
20      140734123              SECONDARY CODES   
21      140734123                      ASSAULT   
22      140734822                LARCENY/THEFT   
23      146178476                LARCENY/THEFT   
24      146179355                LARCENY/THEFT   
25      140734151               OTHER OFFENSES   
26      140734010                      ASSAULT   
27      140734010                  WEAPON LAWS   
28      140734010               OTHER OFFENSES   
29      140736624                VEHICLE THEFT   
...           ...                          ...   
28963   146110064                 NON-CRIMINAL   
28964   146114903                LARCENY/THEFT   
28965   146105009                 NON-CRIMINAL   
28966   146105065                LARCENY/THEFT   
28967   146105504                LARCENY/THEFT   
28968   146105554                LARCENY/THEFT   
28969   140455711                LARCENY/THEFT   
28970   140455711                LARCENY/THEFT   
28971   140455705                     WARRANTS   
28972   140455705                     WARRANTS   
28973   140636701               OTHER OFFENSES   
28974   140456032                     TRESPASS   
28975   150091383                     TRESPASS   
28976   150091383               SUSPICIOUS OCC   
28977   150099149                LARCENY/THEFT   
28978   140605461               OTHER OFFENSES   
28979   140626865                LARCENY/THEFT   
28980   140626865                LARCENY/THEFT   
28981   140691733               OTHER OFFENSES   
28982   140457331                        FRAUD   
28983   140801415               OTHER OFFENSES   
28984   146106938                LARCENY/THEFT   
28985   141085064               OTHER OFFENSES   
28986   146107033                LARCENY/THEFT   
28987   146108495                LARCENY/THEFT   
28988   140763495                 NON-CRIMINAL   
28989   140779301                      ASSAULT   
28990   140908839                LARCENY/THEFT   
28991   140908839                LARCENY/THEFT   
28992   141017556                 NON-CRIMINAL   

                                                Descript DayOfWeek  \
0                                     ARSON OF A VEHICLE    Sunday   
1                                          LOST PROPERTY    Sunday   
2                           GRAND THEFT FROM LOCKED AUTO    Sunday   
3                           GRAND THEFT FROM LOCKED AUTO    Sunday   
4                                         FOUND PROPERTY    Sunday   
5                                POSSESSION OF MARIJUANA    Sunday   
6            POSSESSION OF CONTROLLED SUBSTANCE FOR SALE    Sunday   
7           DRIVING WHILE UNDER THE INFLUENCE OF ALCOHOL    Sunday   
8                    EVADING A POLICE OFFICER RECKLESSLY    Sunday   
9                                            TRESPASSING    Sunday   
10                                          STOLEN TRUCK    Sunday   
11                               PETTY THEFT OF PROPERTY    Sunday

The columns of the dataframe are as follows:

* **IncidntNum** id of the incident (
* **Category** which category the incident
* **Descript** a description of the incident
* **DayOfWeek** on which weekday did the incident happen
* **Date** date of the incident
* **Time** time of the incident
* **PdDistrict**  in which district did the incident happen
* **Resolution** how was the situation resolved
* **Address** the street address where the incident happened
* **X** X coordinate of the incident
* **Y** Y coordinate of the incident
* **Location** X and Y coordinates of the incident combined
* **PdId** another ID number of the incident

Before you start to work in the problems please have a look at the hint at the end of the notebook.

### Problem 8-1:

Think about which parts of the dataset are not useful for out analysis. Is information repeated? If so, you could delete the corresponding columns in order to speed up analysis. You can use the `drop` method (try `help(df.drop)` of the DataFrame `df` to delete one or more columns. In case you want to delete more than one column you need to provide a list of column names.

Get an overview about the dataset and try to answer the following questions. You can double-click on the question your edit your answers:

* How many row are there in the dataset? **28993 rows**
* How seems the data to be ordered? **Date and Time 降序排列**
* About which dates was the date collected? **06/01/2014 - 08/31/2014**

In [6]:
df=df.drop(['X','Y'],axis=1)
df

IncidntNum                     Category  \
0       140734311                        ARSON   
1       140736317                 NON-CRIMINAL   
2       146177923                LARCENY/THEFT   
3       146177531                LARCENY/THEFT   
4       140734220                 NON-CRIMINAL   
5       140734349                DRUG/NARCOTIC   
6       140734349                DRUG/NARCOTIC   
7       140734349  DRIVING UNDER THE INFLUENCE   
8       140738147               OTHER OFFENSES   
9       140734258                     TRESPASS   
10      140738711                VEHICLE THEFT   
11      146178410                LARCENY/THEFT   
12      140787570                 NON-CRIMINAL   
13      140735604                VEHICLE THEFT   
14      140738147                VEHICLE THEFT   
15      140734145                      ASSAULT   
16      140734214                        FRAUD   
17      140734214                      ASSAULT   
18      140734098               SUSPICIOUS OCC   
19      140734098                 NON-CRIMINAL   
20      140734123              SECONDARY CODES   
21      140734123                      ASSAULT   
22      140734822                LARCENY/THEFT   
23      146178476                LARCENY/THEFT   
24      146179355                LARCENY/THEFT   
25      140734151               OTHER OFFENSES   
26      140734010                      ASSAULT   
27      140734010                  WEAPON LAWS   
28      140734010               OTHER OFFENSES   
29      140736624                VEHICLE THEFT   
...           ...                          ...   
28963   146110064                 NON-CRIMINAL   
28964   146114903                LARCENY/THEFT   
28965   146105009                 NON-CRIMINAL   
28966   146105065                LARCENY/THEFT   
28967   146105504                LARCENY/THEFT   
28968   146105554                LARCENY/THEFT   
28969   140455711                LARCENY/THEFT   
28970   140455711                LARCENY/THEFT   
28971   140455705                     WARRANTS   
28972   140455705                     WARRANTS   
28973   140636701               OTHER OFFENSES   
28974   140456032                     TRESPASS   
28975   150091383                     TRESPASS   
28976   150091383               SUSPICIOUS OCC   
28977   150099149                LARCENY/THEFT   
28978   140605461               OTHER OFFENSES   
28979   140626865                LARCENY/THEFT   
28980   140626865                LARCENY/THEFT   
28981   140691733               OTHER OFFENSES   
28982   140457331                        FRAUD   
28983   140801415               OTHER OFFENSES   
28984   146106938                LARCENY/THEFT   
28985   141085064               OTHER OFFENSES   
28986   146107033                LARCENY/THEFT   
28987   146108495                LARCENY/THEFT   
28988   140763495                 NON-CRIMINAL   
28989   140779301                      ASSAULT   
28990   140908839                LARCENY/THEFT   
28991   140908839                LARCENY/THEFT   
28992   141017556                 NON-CRIMINAL   

                                                Descript DayOfWeek  \
0                                     ARSON OF A VEHICLE    Sunday   
1                                          LOST PROPERTY    Sunday   
2                           GRAND THEFT FROM LOCKED AUTO    Sunday   
3                           GRAND THEFT FROM LOCKED AUTO    Sunday   
4                                         FOUND PROPERTY    Sunday   
5                                POSSESSION OF MARIJUANA    Sunday   
6            POSSESSION OF CONTROLLED SUBSTANCE FOR SALE    Sunday   
7           DRIVING WHILE UNDER THE INFLUENCE OF ALCOHOL    Sunday   
8                    EVADING A POLICE OFFICER RECKLESSLY    Sunday   
9                                            TRESPASSING    Sunday   
10                                          STOLEN TRUCK    Sunday   
11                               PETTY THEFT OF PROPERTY    Sunday

### Problem 8-2:

How many `Categories` of crimes are in the statistic. For each category, calculate how many crimes of the category were committed. Provide source code for your answer and summarize your findings below.

In [13]:
df2=df.groupby('Category').size()
print(df2)
len(df2)

Category
ARSON                            63
ASSAULT                        2882
BRIBERY                           1
BURGLARY                          6
DISORDERLY CONDUCT               31
DRIVING UNDER THE INFLUENCE     100
DRUG/NARCOTIC                  1345
DRUNKENNESS                     147
EMBEZZLEMENT                     10
EXTORTION                         7
FAMILY OFFENSES                  10
FORGERY/COUNTERFEITING           18
FRAUD                           242
GAMBLING                          1
KIDNAPPING                      117
LARCENY/THEFT                  9466
LIQUOR LAWS                      42
LOITERING                         3
MISSING PERSON                 1266
NON-CRIMINAL                   3023
OTHER OFFENSES                 3567
PORNOGRAPHY/OBSCENE MAT           1
PROSTITUTION                    112
ROBBERY                         308
RUNAWAY                          61
SECONDARY CODES                 442
STOLEN PROPERTY                   8
SUICIDE            

34

**34种犯罪类型**

### Problem 8-3:

How many incidents happened on each day of the week?

In [27]:
df3=df.groupby('DayOfWeek').size()
df3

DayOfWeek
Friday       4451
Monday       4005
Saturday     4319
Sunday       4218
Thursday     3968
Tuesday      3930
Wednesday    4102
dtype: int64

**周末以及邻近周末犯罪数量比较多**

### Problem 8-4:

On which day were the most people arrested.

In [51]:
df4=df.groupby('Date').size()
df4.sort_values(ascending=False)

Date
08/09/2014    410
06/28/2014    410
08/08/2014    403
06/29/2014    397
08/29/2014    388
06/04/2014    380
08/23/2014    373
08/01/2014    372
08/22/2014    371
06/20/2014    369
08/16/2014    364
08/10/2014    359
06/27/2014    356
08/07/2014    351
07/21/2014    350
07/22/2014    348
08/18/2014    345
07/24/2014    344
07/07/2014    343
08/14/2014    342
07/29/2014    342
08/02/2014    340
07/27/2014    339
07/09/2014    338
08/30/2014    337
08/26/2014    335
08/20/2014    334
08/13/2014    330
07/28/2014    330
07/25/2014    328
             ... 
06/17/2014    293
07/13/2014    292
06/13/2014    292
07/17/2014    291
07/10/2014    290
07/05/2014    290
07/15/2014    290
08/03/2014    289
06/10/2014    289
07/20/2014    288
08/21/2014    287
07/26/2014    285
06/01/2014    284
06/14/2014    283
06/30/2014    281
08/28/2014    280
08/05/2014    279
06/12/2014    279
07/08/2014    278
06/19/2014    277
06/15/2014    275
06/11/2014    275
08/04/2014    274
06/16/2014    274
08/19

**06/28/2014 & 08/09/2014，count=410**

### Problem 8-5:

For each crime category calculate how many incidents of this type were committed on a workday (Monday--Friday) and on the weekend (Saturday, Sunday). Are some incidents more prominent on the weekend?

In [54]:
def f(DayOfWeek):
    if DayOfWeek in ['Saturday','Sunday']:
        return 'weekend'
    return 'workday'
df['DayType']=df.DayOfWeek.apply(f)
df5=df.groupby(['Category','DayType']).size()
df5
#df5.sort_values(ascending=False)

Category                     DayType
ARSON                        weekend      22
                             workday      41
ASSAULT                      weekend     914
                             workday    1968
BRIBERY                      workday       1
BURGLARY                     weekend       1
                             workday       5
DISORDERLY CONDUCT           weekend       9
                             workday      22
DRIVING UNDER THE INFLUENCE  weekend      42
                             workday      58
DRUG/NARCOTIC                weekend     301
                             workday    1044
DRUNKENNESS                  weekend      50
                             workday      97
EMBEZZLEMENT                 weekend       3
                             workday       7
EXTORTION                    weekend       2
                             workday       5
FAMILY OFFENSES              workday      10
FORGERY/COUNTERFEITING       weekend       4
                  

**Your answer here**

### Problem 8-6:

Calculate how many incidents happened of the day (between 00 and 23). When did the most incidents occur. 

In [57]:
df6=df.groupby(['Time']).size()
df6.sort_values(ascending=False)

Time
12:00    784
00:01    661
18:00    649
19:00    621
17:00    594
20:00    586
22:00    554
15:00    507
21:00    503
16:00    475
23:00    467
13:00    447
14:00    432
10:00    394
09:00    378
08:00    366
11:00    346
18:30    334
19:30    302
16:30    287
17:30    282
22:30    274
15:30    271
20:30    263
21:30    258
01:00    245
23:30    236
12:30    213
11:30    210
14:30    209
        ... 
04:36      1
01:26      1
04:31      1
04:04      1
04:09      1
04:17      1
04:16      1
04:14      1
14:46      1
05:03      1
03:37      1
05:07      1
06:03      1
06:02      1
04:11      1
05:54      1
03:23      1
05:48      1
05:39      1
05:37      1
05:31      1
00:58      1
05:27      1
05:26      1
05:24      1
07:37      1
03:31      1
05:13      1
05:11      1
07:38      1
dtype: int64

**12:00发生最多**

### Problem 8-7:
Compare the number of incidents for each hour on weekdays and on weekends. Do the most number of incidents happen on different hours on weekends and weekdays?

**Your answer here**

### Problem 8-8:

Were some incidents more prominent on certain hours? Calculate how many incidents of each type happened for each hour of the day.

**Your answer here**

### Problem 8-9:

How many police districts are there in San Francisco? Calculate the number of crimes in each district? Can you deduce that some areas are more dangerous than others?

In [ ]:
df.groupby('PdDistrict').size()

**Your answer here**

### Problem 8-10:

The following are the population numbers of the 10 police districts. Contrast them with them number of crimes in each district. What can you say about the crime rate (crimes/population) of each district? Should you consider all incidents are only some?

```
Bayview 60,301
Central 69,276
Ingleside 132,328
Mission 83,235
Northern 82,348
Park 59,572
Richmond 93,693
Southern 24,157
Taraval 147,806
Tenderloin 21,669
```
(data from the year 2000, see here https://sanfranciscopolice.org/sites/default/files/FileCenter/Documents/14683-SFPD_DSBAfinal_trnsmtl.pdf)

**Your answer here**

## Pivoting tables

Pivot tables are useful for some of the problems. We will talk about them next Monday, but here is a little example from the pandas documentation (http://pandas.pydata.org/pandas-docs/stable/reshaping.html). We first create a dataframe, you **do not need to understand** this cell, just execute it and have a look at the resulting dataframe:

In [ ]:
# We first create a dataframe, you do not need to understand this function:
import pandas.util.testing as tm; tm.N = 3
import numpy as np
def unpivot(frame):
    N, K = frame.shape
    data = {'value' : frame.values.ravel('F'),
            'variable' : np.asarray(frame.columns).repeat(N),
            'date' : np.tile(np.asarray(frame.index), K)}
    return pd.DataFrame(data, columns=['date', 'variable', 'value'])
df_hint = unpivot(tm.makeTimeDataFrame())
df_hint

Observe that we have 12 rows of data, one for each combination of **date** and **variable** values. The data in the table would be easier to understand if we could use the values of **date** and **variable** as labels. The `pivot` method does just that:

In [ ]:
df_hint.pivot(index='date', columns='variable', values='value')

We could also *pivot in the other direction*, using **date** for the columns and **variable** for the rows:

In [ ]:
df_hint.pivot(index='variable', columns='date', values='value')

Another option is the `pivot_table` method, which combines the functionality of `groupby` and `pivot`:

In [ ]:
df_pt = pd.DataFrame({'name': list("ABCDBADC" * 2), 'major': ('java,math,' *8).split(',')[:16],
                      'scores': list(range(40, 90, 3))[:16]})
df_pt

In [ ]:
pd.pivot_table(df_pt, columns='major', index='name', values='scores', aggfunc=np.max)

Observe that in each field we now have the maximum value of the two scores. Possible values for `aggfunc` are `np.max`, `np.min`, `np.sum`, `np.mean`, and many more. We can even combine them as follows:

In [ ]:
pd.pivot_table(df_pt, columns='major', index='name', values='scores', 
               aggfunc=[np.max, np.mean])